Add FLASHIAP component to DISCO_H747

JanneKiiskila opened the pull request:
### Description

FLASHIAP component/capability is needed for DeviceKey functionality
and also for Pelion enablement - firmware update uses this feature, too.

Related to issue [11617](https://github.com/ARMmbed/mbed-os/issues/11617).


    [ ] Fix
    [ ] Refactor
    [x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@ARMmbed/team-st-mcd  @jeromecoutant @adbridge 

### Release Notes

Enable FLASHIAP component on ST DISCO_H747I.

Commit:
Add FLASHIAP component to DISCO_H747

FLASHIAP component/capability is needed for DeviceKey functionality
and also for Pelion enablement - firmware update uses this feature, too.

jeromecoutant commented on targets/targets.json:
```diff
@@ -3364,7 +3365,7 @@
         ],
         "release_versions": ["2", "5"],
-        "bootloader_supported": true
+        "bootloader_supported": false
```
not related to the PR ?

JanneKiiskila replied:
Good catch, definitely not. That's an early bring-up change to avoid the `MBED_ROM_START` problems. I'll force push a fix.

jeromecoutant reviewed (commented):
| target            | platform_name | test suite                                               | result | elapsed_time (sec) | copy_method |
|-------------------|---------------|----------------------------------------------------------|--------|--------------------|-------------|
| DISCO_H747I-ARMC6 | DISCO_H747I   | tests-mbed_drivers-flashiap                              | OK     | 38.6               | default     |
| DISCO_H747I-ARMC6 | DISCO_H747I   | features-device_key-tests-device_key-functionality          | FAIL   | 44.45              | default     |
| DISCO_H747I-ARMC6 | DISCO_H747I   | features-storage-tests-kvstore-direct_access_devicekey_test | FAIL   | 28.1               | default     |

Commit:
DISCO_H747I - define default storage (TDB_INTERNAL)

DeviceKey needs the definition of the default storage place,
define it to be TDB_INTERNAL (as for the other boards).

jeromecoutant commented on features/storage/kvstore/conf/global/mbed_lib.json:
```diff
@@ -20,6 +20,9 @@
             "storage_type": "TDB_INTERNAL"
         },
+        "DISCO_H747I": {
+            "storage_type": "TDB_INTERNAL"
```
Maybe it will be easier to set the default value to TDB_INTERNAL ?
instead of overriding each target one by one ?

JanneKiiskila replied:
@SeppoTakalo  - we should consider this?

SeppoTakalo replied:
Yes, apparently many modules are assuming that KVStore "just works", so we need to probably default to TDB_INTERNAL and silently steal some flash away from the app.

But of course it needs to be configurable, so that location can be overwritten.

Commit:
Add default TDBStore location for DISCO_H747I (end of BANK1)

Place it at the end of the bank1, last two erase sectors.
As erase sector is 128 kB, default size must be double of that.

This can't be in bank2 as that is apparently dedicated to the M4 core.

Memory map is available in;

https://github.com/ARMmbed/mbed-os/blob/master/targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/flash_data.h

Memory map does not have this information, but issue [11617](https://github.com/ARMmbed/mbed-os/issues/11617) has.

jeromecoutant commented on features/storage/kvstore/conf/tdb_internal/mbed_lib.json:
```diff
@@ -15,6 +15,10 @@
             "internal_size": "0x8000",
             "internal_base_address": "0x00028000"
         },
+        "DISCO_H747": {
```
Wrong target name :-(

JanneKiiskila replied:
Darn diddly, good catch. It's fixed now.

Commit:
Fix for the H747 flash driver / cache cleaning

This copies the approach of the STM32F7 flash driver submitted via
PR https://github.com/ARMmbed/mbed-os/pull/10248

With this change the board finally passes all of the device key
tests 10/10 times correctly.

LMESTM commented on features/storage/kvstore/conf/tdb_internal/mbed_lib.json:
```diff
@@ -15,6 +15,10 @@
         },
+        "DISCO_H747I": {
+            "internal_size": "256*1024",
+            "internal_base_address": "0x080C0000"
```
I don't like so much having this value HARD CODED. 
Couldn't this be an address like 
" MBED_APP_START + MBED_APP_SIZE - 256*1024 + 1 " 
?
Note: I'm okay to move on wit this first implementation and try to make a later modification that makes it more dynamic ...

JanneKiiskila replied:
I wonder wha the MBED_APP_SIZE is actually here? If it's 1024-(2*128), then it would be OK. We can do that change for sure.

JanneKiiskila replied:
But at least the offsets for example, you can't give as computed things - they had to be in real numbers.

LMESTM replied:
The offset you're referring should be "internal_size" right ? Basically we're allocating a region of size "internal_size" at the end of application flash memory, isn't it ? so if we get the end address, we can deduce the rest of it at least for default value. Of course application can overload to somewhere else in memory and over-write the 2 parameters.

LMESTM reviewed (approved):
LGTM - with 1 comment that could be addressed at a later stage
